### PR TITLE
Support to conditions, filters (transformers/modifiers) and Closure Values

### PR DESCRIPTION
**Pull request by rahal**
Hello, first of all, thank you for this lightweight template library, I needed some more features as I plan to use it in my wordpress plugin ["Mail Control"](https://fr.wordpress.org/plugins/mail-control/) .

Would be happy to get your feedback about them : 

- Add support to simple conditions :
	- {#name}if true{/name}
	- {#name}if true{#else}if false{/name}
	- {#name}if true {variable}{#else}if false{other_variable}{/name}
- Add support to simple filters ( transformers/modifiers ) :
	- {variable|upper} , {variable|lower}, {variable|esc_html}
	- Allow defining additionnal/override filters in constructor
- Add support to Closure Values
- Updated Readme and composer file (bench command)

### CODE REVIEW

**Comment by marios88** [comment]
I have done something similar with a fork.

IMO the base StringTemplate should remain as simple as it without any fluff.

I really do like the closure tho !

**Comment by rahal** [comment]
Hi @marios88,
Thank you for your feedback
> I have done something similar with a fork.

I checked pull requests and other forks before adding those features, I just recall seeing this commit : https://github.com/nicmart/StringTemplate/commit/327748460a0d0267c8bc1099da9f62a17666b1d5 ( that inspired me to add the Closure support as I found it interesting and simple enough )

> IMO the base StringTemplate should remain as simple as it without any fluff.

As I said, I needed the conditional feature (and filters too), and I couldn't find a more lightweight library than this one ( I considered  https://github.com/bobthecow/mustache.php , but I like to keep my code to the minimum ).
That said, I updated and ran the benchmark and this new implementation (even if it has more features is faster than SprintfEngine ) : `https://github.com/Youniwemi/StringTemplate/actions/runs/4385753705/jobs/7678887580`

Sure, we can't do simpler than str_replace, but in my use case, most of the time, I'll be sending probably more values to the render function than the variables to be replaced, so it seems to me like it's not really efficient to loop over all the values to replace them.

That said, as this implementation doesn't need NestedKeyIterator and RecursiveArrayOnlyIterator, they could be removed (but that's an other story as people maybe using them) 

> I really do like the closure tho !

Thank You again for your feedback, but I'm not sure what you are suggesting, do you think I should create a new "Engine"  ?

**Comment by marios88** [comment]
> Thank You again for your feedback, but I'm not sure what you are suggesting, do you think I should create a new "Engine" ?

I just extended in my own repo and as i needed an easy and fast way to translate strings and hardcoded "{(...)}" to go through translation

The closure addition would be nice for the core StringTemplate IMO, but the other features remind me of more full blown engines. Why not use Twig, latte etc?

**Comment by rahal** [comment]
> I just extended in my own repo and as i needed an easy and fast way to translate strings and hardcoded "{(...)}" to go through translation

Nice syntax! with my implementation,  you could have passed a custom filter "translate"  or "t" in the constructor, and used it like this  {variable|t}

> The closure addition would be nice for the core StringTemplate IMO, but the other features remind me of more full blown engines. Why not use Twig, latte etc?

Why not twig? Mustache or Latte? Too much code to be added as a dependency in a small wordpress plugin (which will probably coexist with other plugins), a little overkill I think. See for yourself ( Lines of code, I used this url for the count https://codetabs.com/count-loc/count-loc-online.html , I think the count includes test code)

Mustache.php : 6792
Latte : 15293
twig : 24734
nicmast/StringTemplate: 751
Youniwemi/StringTemplate Fork : 1098  ( but can be reduced as  NestedKeyIterator and RecursiveArrayOnlyIterator and SprintfEngine  (slower and less functionnal than the new Engine), AbstractEngine are not longer needed - except for backward compatibility)

The bottom line, I needed just a little more than what this library offers (an much less than what other full blown engines but if I had to pick, I would pick mustache :)  I somehow got inspired from their syntax for the conditionnal.

**Comment by rahal** [comment]
Hello,

I hope this message finds you well. I submitted a pull request to this project some time ago, but it appears that the project's maintainers prefer to keep it in its current, simpler form.

I want to express my appreciation for your work on this project and its simplicity, which I found appealing due to the reduced lines of code. However, I have decided to close the pull request as I intend to maintain a fork of the project.

I'd like to invite you to collaborate with me on this forked project, as I believe our combined efforts could lead to further improvements. I'm already using this work in two free WordPress plugins, Mail Control ([link](https://wordpress.org/plugins/mail-control/)) and AIIFY ([link](https://wordpress.org/plugins/aiify/)), as well as in Instareza ([link](https://www.instareza.com)).

Please feel free to reach out if you have any concerns or feedback. I value your input and would appreciate the opportunity to work together.

Thank you again for your work on this project, and I look forward to the possibility of collaboration.
